libraries/chain/db_market.cpp

```diff
@@ -62,34 +62,61 @@ void database::globally_settle_asset( const asset_object& mia, const price& sett
    const asset_dynamic_data_object& mia_dyn = mia.dynamic_asset_data_id(*this);
    auto original_mia_supply = mia_dyn.current_supply;
 
-   const auto& call_price_index = get_index_type<call_order_index>().indices().get<by_price>();
-
    auto maint_time = get_dynamic_global_properties().next_maintenance_time;
    bool before_core_hardfork_342 = ( maint_time <= HARDFORK_CORE_342_TIME ); // better rounding
+   bool before_core_hardfork_1270 = ( maint_time <= HARDFORK_CORE_1270_TIME ); // call price caching issue
 
    // cancel all call orders and accumulate it into collateral_gathered
-   auto call_itr = call_price_index.lower_bound( price::min( bitasset.options.short_backing_asset, mia.id ) );
-   auto call_end = call_price_index.upper_bound( price::max( bitasset.options.short_backing_asset, mia.id ) );
+   const auto& call_index = get_index_type<call_order_index>().indices();
+   const auto& call_price_index = call_index.get<by_price>();
+   const auto& call_collateral_index = call_index.get<by_collateral>();
+
+   auto call_min = price::min( bitasset.options.short_backing_asset, mia.id );
+   auto call_max = price::max( bitasset.options.short_backing_asset, mia.id );
+
+   auto call_price_itr = call_price_index.begin();
+   auto call_price_end = call_price_itr;
+   auto call_collateral_itr = call_collateral_index.begin();
+   auto call_collateral_end = call_collateral_itr;
+
+   if( before_core_hardfork_1270 )
```

Comment:
I think the distinction before/after hardfork can be implemented more cleanly by moving most of this method into a templated function with the index as a template parameter.

Reply:
Correct. I almost forgot how to write a template function. Will try.

Reply:
Done: https://github.com/bitshares/bitshares-core/pull/1672/commits/6faa519351f55a5440d649f1d97394123e062318.